plugins/UM3NetworkPrinting/src/SendMaterialJob.py

```diff
@@ -1,99 +1,129 @@
             Logger.log("e", "Request material storage on printer: I didn't understand the printer's answer.")
             return
-        try:
-            remote_materials_by_guid = {material["guid"]: material for material in remote_materials_list} #Index by GUID.
         except KeyError:
             Logger.log("e", "Request material storage on printer: Printer's answer was missing GUIDs.")
             return
 
-        container_registry = ContainerRegistry.getInstance()
-        local_materials_list = filter(lambda material: ("GUID" in material and "version" in material and "id" in material), container_registry.findContainersMetadata(type = "material"))
-        local_materials_by_guid = {material["GUID"]: material for material in local_materials_list if material["id"] == material["base_file"]}
-        for material in local_materials_list: #For each GUID get the material with the highest version number.
-            try:
-                if int(material["version"]) > local_materials_by_guid[material["GUID"]]["version"]:
-                    local_materials_by_guid[material["GUID"]] = material
-            except ValueError:
-                Logger.log("e", "Material {material_id} has invalid version number {number}.".format(material_id = material["id"], number = material["version"]))
-                continue
+        # Collect local materials
+        local_materials_by_guid = self._getLocalMaterials()
 
-        materials_to_send = set() #type: Set[Dict[str, Any]]
-        for guid, material in local_materials_by_guid.items():
-            if guid not in remote_materials_by_guid:
-                materials_to_send.add(material["id"])
-                continue
-            try:
-                if int(material["version"]) > remote_materials_by_guid[guid]["version"]:
-                    materials_to_send.add(material["id"])
-                    continue
-            except KeyError:
-                Logger.log("e", "Current material storage on printer was an invalid reply (missing version).")
-                return
+        # Find out what materials are new or updated annd must be sent to the printer
+        materials_to_send = {
+            material.id
+            for guid, material in local_materials_by_guid.items()
+            if guid not in remote_materials_by_guid or
+               material.version > remote_materials_by_guid[guid].version
+        }
 
+        # Send materials to the printer
+        self.sendMaterialsToPrinter(materials_to_send)
+
+    def sendMaterialsToPrinter(self, materials_to_send):
         for file_path in Resources.getAllResourcesOfType(CuraApplication.ResourceTypes.MaterialInstanceContainer):
             try:
                 mime_type = MimeTypeDatabase.getMimeTypeForFile(file_path)
             except MimeTypeDatabase.MimeTypeNotFoundError:
-                continue #Not the sort of file we'd like to send then.
+                continue  # Not the sort of file we'd like to send then.
+
```

Comment:
It might be nice to extract a method for the actual sending so there is not a giant code block in this for loop.

Reply:
I tried to split it up into two parts, but it does not make it clearer. The file_name used in the extraction of the material_id is also used to send the contents of the file to the printer.

Reply:
It would make it easier to mock though. Right now you need a ton of mocking & patching to fix this (which makes the test pretty hard to read)